pkg/resource/deploy/source_query.go

```diff
@@ -0,0 +1,274 @@
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to start resource monitor")
+	}
+
+	// Create a new iterator with appropriate channels, and gear up to go!
+	src := &querySource{
+		mon:     mon,
+		plugctx: plugctx,
+		runinfo: runinfo,
+		finChan: make(chan result.Result),
+		cancel:  cancel,
+	}
+
+	// Now invoke Run in a goroutine.  All subsequent resource creation events will come in over the gRPC channel,
+	// and we will pump them through the channel.  If the Run call ultimately fails, we need to propagate the error.
+	src.forkRun()
+
+	// Finally, return the fresh iterator that the caller can use to take things from here.
+	return src, nil
+}
+
+type querySource struct {
+	mon     *queryResmon       // the resource monitor, per iterator.
+	plugctx *plugin.Context    // the plugin context.
+	runinfo *EvalRunInfo       // the directives to use when running the program.
+	finChan chan result.Result // the channel that communicates completion.
+	done    bool               // set to true when the evaluation is done.
+	cancel  context.Context
+}
+
+func (src *querySource) Close() error {
+	// Cancel the monitor and reclaim any associated resources.
+	src.done = true
+	return src.mon.Cancel()
+}
+
+func (src *querySource) Wait() result.Result {
+	// If we are done, quit.
+	if src.done {
+		return nil
+	}
+
+	select {
+	case res := <-src.finChan:
+		src.Close()
+		return res
+	case <-src.cancel.Done():
+		src.Close()
+		return nil
+	}
+}
+
+// forkRun evaluate the query program in a separate goroutine. Completion or cancellation will cause
+// `Wait` to stop blocking and return.
+func (src *querySource) forkRun() {
+	// Fire up the goroutine to make the RPC invocation against the language runtime.  As this executes, calls
+	// to queue things up in the resource channel will occur, and we will serve them concurrently.
+	go func() {
+		// Next, launch the language plugin.
+		run := func() result.Result {
+			rt := src.runinfo.Proj.Runtime.Name()
+			langhost, err := src.plugctx.Host.LanguageRuntime(rt)
+			if err != nil {
+				return result.FromError(errors.Wrapf(err, "failed to launch language host %s", rt))
+			}
+			contract.Assertf(langhost != nil, "expected non-nil language host %s", rt)
+
+			// Make sure to clean up before exiting.
+			defer contract.IgnoreClose(langhost)
+
+			// Decrypt the configuration.
+			config, err := src.runinfo.Target.Config.Decrypt(src.runinfo.Target.Decrypter)
+			if err != nil {
+				return result.FromError(err)
+			}
+
+			// Now run the actual program.
+			progerr, bail, err := langhost.Run(plugin.RunInfo{
+				MonitorAddress:   src.mon.Address(),
+				Stack:            string(src.runinfo.Target.Name),
+				Project:          string(src.runinfo.Proj.Name),
+				Pwd:              src.runinfo.Pwd,
+				Program:          src.runinfo.Program,
+				Args:             src.runinfo.Args,
+				Config:           config,
+				DryRun:           true,
+				QueryModeEnabled: true,
```

Comment:
We don't use the `Enabled` postfix for other booleans.  Should we name this just `QueryMode` or `Query`?

Reply:
We did, I thought, for test mode.

Reply:
Not in the same way - only for some internal helpers.  I'd stay consistent with the approach used in the siblings here (and elsewhere this is passed through).